CoroutineTemplate/domain/src/main/java/co/nimblehq/coroutine/usecase/GetUsersUseCase.kt

```diff
@@ -0,0 +1,12 @@
+package co.nimblehq.coroutine.usecase
+
+import co.nimblehq.coroutine.entity.UserEntity
+import co.nimblehq.coroutine.repository.UserRepository
+import javax.inject.Inject
+
+class GetUsersUseCase @Inject constructor(private val userRepository: UserRepository) {
+
+    suspend fun execute(): UseCaseResult<List<UserEntity>> {
+        return userRepository.getUsers()
```

Comment:
shouldn't you catch the exception here?

Reply:
@sleepylee No need, we have `try/catch` in `RepositoryImpl` already to make sure the `Repository` always returns UseCaseResult.

Reply:
erh....why would Repository know about the UseCase?

Reply:
@sleepylee Ah the `UseCaseResult` is just a `Model Class` that contains Result or Error.
Certainly, the Repository would not know about the UseCase.

Reply:
@doannimble there is a flaw in this design, combining with the comment I put above:
- The UseCaseResult.Error is very useless atm, it doesn't contain a useful Exception type for further handling.
- The Repository acts as the source of connecting various data type, while it can leverage multiple sources, it doesn't have to tie the result to the downstream classes (UseCase) that are actually using it.

I'm not sure why wouldn't we just through the Exception to the UseCase, capturing it and returns as UseCaseResult.Error or .Success? In the end, the UseCase expects to returns `UseCaseResult` from the `call()` method right?

Reply:
@doannimble `RepositoryImpl` should catch exceptions related to I/O such as network, database, etc. and map them to some common errors (e.g., NoConnectionError, ApiHttpError, etc.) recognized by the UseCase. It's up to the UseCase to catch those errors and map to more specific errors that would be consumed by the `presentation` layer, e.g., `DefaultUserError`, `CreateUserError`, etc.

Reply:
@AVI5HEK Yes, I'm doing like that, `RepositoryImpl` just catch exceptions related to Network so far.
Hence I think, the Error mapper should happen on `UseCase`